chore: release-please action

vlnevyhosteny opened the pull request:
Release-As: 2.0.1

ViolanteCodes commented:
@vlnevyhosteny - I've assigned my colleague, @Toreno96 to this review, as he has more experience with github actions and has looked at the other similar PRs; however, can you please update the commit message from "chore" to "feat"? "chore" will not generate a new release with the `release-please` action.

Toreno96 commented on .github/workflows/release-please.yml:
```diff
@@ -0,0 +1,27 @@
+# This workflow opens and updates a pull request with a new package version
+# based on code changes.
```
Could you expand those comments in a similar manner as https://github.com/ButterCMS/buttercms-ruby/pull/28/files#diff-2c84033033d49186c63e6adcd705f63b11ae6814cd76c152c9c486d389fbccf3R4-R7?

vlnevyhosteny replied:
@Toreno96 Resolved in https://github.com/ButterCMS/buttercms-csharp/pull/40/commits/f2e4a865ea97785f7146bd83c4808969a15a6fd4

Toreno96 commented on .github/workflows/release-please.yml:
```diff
@@ -0,0 +1,27 @@
+    steps:
+      - uses: google-github-actions/release-please-action@v3.7.10
+        with:
+          release-type: simple
```
According to the https://github.com/marketplace/actions/release-please-action#release-types-supported, `simple` release type requires `version.txt` and `CHANGELOG.md`, but I don't see the former in the repo:

![2023-10-04_16-14-40_CEST_screenshot](https://github.com/ButterCMS/buttercms-csharp/assets/16722326/d583f8e7-ee7e-4895-85eb-58673d87a99c)

Are you sure it will work as expected?

vlnevyhosteny replied:
@Toreno96  I use the release type "simple" because release-please lib does not support .NET yet. But it works since I included Release-As: 2.0.0 in the commit message to initialize versioning (it won't start from 1.0.0.). After merging this release, it iterates over versions as expected. I have tested that on my fork: 

- This is PR with releasing 2.0.0 version - https://github.com/vlnevyhosteny/buttercms-csharp/pull/6 . As you can see, it just generates a changelog.
- Then I tried to create some dummy changes to test it, and the correct PR was created - https://github.com/vlnevyhosteny/buttercms-csharp/pull/5 . As it was a feat, it iterated from 2.0.0 to 2.1.0.

So I can include version.txt, but it's not necessary.

Toreno96 replied:
OK, I was actually curious what's that `Release-As: 2.0.0` in the PR description, didn't know it was for `release-please`. Nice trick!

Thanks for the detailed test.

> So I can include version.txt, but it's not necessary.

Not necessary indeed, let's leave it as is.

vlnevyhosteny commented:
> @vlnevyhosteny - I've assigned my colleague, @Toreno96 to this review, as he has more experience with github actions and has looked at the other similar PRs; however, can you please update the commit message from "chore" to "feat" or "fix"? "chore" will not generate a new release with the `release-please` action.

@ViolanteCodes Ok, I will rebase commit to be "feat". However I think it's not really correct since these changes does not impact users that using SDK. It's more infrastructure related.

Toreno96 commented:
> @ViolanteCodes Ok, I will rebase commit to be "feat". However I think it's not really correct since these changes does not impact users that are using SDK. It's more infrastructure related.

@vlnevyhosteny Can we somehow force `release-please` to deploy the `chore` change?

@ViolanteCodes alternatively, we could consider merging this as `chore`, skipping the deploy, and waiting until we have a change that actually should change the version.
The disadvantage of that solution is that we won't be able to test it all works right away; the advantage is it won't update the version for something that does not impact the user as @vlnevyhosteny mentioned.

Toreno96 commented:
@vlnevyhosteny One other concern I have—shouldn't we add an action (either the separate one or as a part of `release-please.yml`) that publishes the package to NuGet packages as instructed [here](https://learn.microsoft.com/en-us/nuget/nuget-org/publish-a-package)? It seems to me that after this PR, it will only bump the version, but not publish it anywhere, correct?

ViolanteCodes commented:
> @ViolanteCodes alternatively, we could consider merging this as `chore`, skipping the deploy, and waiting until we have a change that actually should change the version. The disadvantage of that solution is that we won't be able to test it all works right away; the advantage is it won't update the version for something that does not impact the user as @vlnevyhosteny mentioned.

I am okay with this, as long as @vlnevyhosteny comes back if the deploy to nuget doesn't work 😉

ViolanteCodes commented:
> @vlnevyhosteny One other concern I have—shouldn't we add an action (either the separate one or as a part of `release-please.yml`) that publishes the package to NuGet packages as instructed [here](https://learn.microsoft.com/en-us/nuget/nuget-org/publish-a-package)? It seems to me that after this PR, it will only bump the version, but not publish it anywhere, correct?

I agree with this - it's what we have for all the other SDK updates - they not only autoversion, but they then deploy out to whatever the package hosting service is

vlnevyhosteny commented:
> I agree with this - it's what we have for all the other SDK updates - they not only autoversion, but they then deploy out to whatever the package hosting service is

@ViolanteCodes @Toreno96  Hmm, I had info from @prokopsimek that for now just release is on the table not publish. 😅  Let me invite him to the discussion to clarify, please.

prokopsimek commented:
I meant to be aware that all the publish actions will differ per a tech stack and only the release-please job will be probably the same. Sorry for confusion. Definitely add both jobs please. 👍😀

On October 5, 2023 at 16:59:08, vnevyhosteny ***@***.*** ***@***.***)) wrote:

> >
> > I agree with this - it's what we have for all the other SDK updates - they not only autoversion, but they then deploy out to whatever the package hosting service is
> >
>
>
> @ViolanteCodes (https://github.com/ViolanteCodes) @Toreno96 (https://github.com/Toreno96) Hmm, I had info from @prokopsimek (https://github.com/prokopsimek) that for now just release is on the table not publish. 😅 Let me invite him to the discussion to clarify, please.
>
>
> —
> Reply to this email directly, view it on GitHub (https://github.com/ButterCMS/buttercms-csharp/pull/40#issuecomment-1749082241), or unsubscribe (https://github.com/notifications/unsubscribe-auth/ABJ3U4M65MTHFCFFWD3C3LTX53DLZAVCNFSM6AAAAAA5QRVOM2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONBZGA4DEMRUGE).
> You are receiving this because you were mentioned.Message ID: ***@***.***>
>
>
>

vlnevyhosteny commented:
> I meant to be aware that all the publish actions will differ per a tech stack and only the release-please job will be probably the same. Sorry for confusion. Definitely add both jobs please. 👍😀

Ahh, my bad sorry. I will add that job then and let you know guys. 🙏🏻

vlnevyhosteny commented:
@ViolanteCodes @Toreno96 So, in https://github.com/ButterCMS/buttercms-csharp/pull/40/commits/f2e4a865ea97785f7146bd83c4808969a15a6fd4 I have added workflow with publish to NuGET. 

I have tested that on https://github.com/vlnevyhosteny/buttercms-csharp/tree/publish-testing and successfully published https://www.nuget.org/packages/Vlads.ButterCMS/ .

Just few things to mention:

- [ ] ⚠️ We have to set up `NUGET_APIKEY` secret before this PR is going to be merged. 
- I added `--skip-duplicate` flag, so publish won't fail after merging this.
- I added job for validating NuGET package right before publish. However, I had to skip this job since there are some error during validation: 

```
Run meziantou.validate-nuget-package $(ls /home/runner/work/buttercms-csharp/buttercms-csharp/nuget/*.nupkg | head -n 1)
{
  "Packages": {
    "/home/runner/work/buttercms-csharp/buttercms-csharp/nuget/Vlads.ButterCMS.2.0.0.nupkg": {
      "Errors": [
        {
          "ErrorCode": 101,
          "Message": "XML documentation not found",
          "HelpText": "lib/netstandard2.1/ButterCMS.dll"
        },
        {
          "ErrorCode": 111,
          "Message": "Symbol file not found",
          "FileName": "lib/netstandard2.1/ButterCMS.dll"
        },
        {
          "ErrorCode": 74,
          "Message": "Repository commit is not set"
        },
        {
          "ErrorCode": 73,
          "Message": "Repository URL is not set"
        },
        {
          "ErrorCode": 72,
          "Message": "Repository type is not set"
        },
        {
          "ErrorCode": 61,
          "Message": "Readme is not set"
        },
        {
          "ErrorCode": 51,
          "Message": "Project url is not set"
        },
        {
          "ErrorCode": 22,
          "Message": "The nuspec file use the deprecated licenseUrl metadata"
        },
        {
          "ErrorCode": 31,
          "Message": "The nuspec file use the deprecated iconUrl metadata"
        }
      ]
    }
  }
}
```

I can tackle those if you want. 🙂

Toreno96 commented:
> I added `--skip-duplicate` flag, so publish won't fail after merging this.

@vlnevyhosteny Is that required because we have already published version `2.0.0` (when upgrading to .NET 2.1) and now we're publishing version `2.0.0` again (with `release-please`-related changes)?

ViolanteCodes commented:
@vlnevyhosteny 

> * [ ]  ⚠️ We have to set up `NUGET_APIKEY` secret before this PR is going to be merged.

Okay - I'll ping Jake on this, as he usually handles the service keys.

> * I added job for validating NuGET package right before publish. However, I had to skip this job since there are some error during validation:

> I can tackle those if you want. 🙂

Yes, please fix. And let us know on Daniels' comment above re: 

> @vlnevyhosteny Is that required because we have already published version `2.0.0` (when upgrading to .NET 2.1) and now we're publishing version `2.0.0` again (with `release-please`-related changes)?

vlnevyhosteny commented:
> > I added `--skip-duplicate` flag, so publish won't fail after merging this.
> 
> @vlnevyhosteny Is that required because we have already published version `2.0.0` (when upgrading to .NET 2.1) and now we're publishing version `2.0.0` again (with `release-please`-related changes)?

@Toreno96 I depends. It's not necessery for the future releases. Unless someone manually trigger publish workflow again. But in this case we are technicaly re-releasing 2.0.0 again so it will fail. There are 3 options:

1. Keep `--skip-duplicate`. The workflow won't fail. It's not going to publish to NuGET again, but it will silently pass. 
2. Remove `--skip-duplicate` so the workflow is going to fail.
3. Remove `--skip-duplicate` and increment patch version (2.0.1). So there will be a new release without any impact on users.

As I'm thinking about that, option 3 makes sense the most.

ViolanteCodes commented:
> As I'm thinking about that, option 3 makes sense the most.

I am fine with this. Releasing a 2.0.1 would also let us test out the SDK push to NuGet and just make sure everything is working properly. Same as before then - we need this commits to not be `chore` so that they'll actually open a version release.

vlnevyhosteny commented:
@ViolanteCodes @Toreno96 So, all validation errors are fixed in https://github.com/ButterCMS/buttercms-csharp/pull/40/commits/013ed27cd9352dc0d72c8a82b6048f71703d2e16. There will be full NuGET listing like this https://www.nuget.org/packages/Vladss.ButterCMS/2.0.1#readme-body-tab . I also edited https://github.com/ButterCMS/buttercms-csharp/pull/40/commits/eadea9adbaa09bcc42ea9e409e2ebbfed2b6efb2 to invoke 2.0.1 release and removed `--skip-duplicate` flag. I think we are ready to go when NUGET_APIKEY secret is provided to the repo 🙂

Toreno96 commented on .github/workflows/nuget-publish.yml:
```diff
@@ -0,0 +1,70 @@
+      - name: Setup .NET Core
+        uses: actions/setup-dotnet@v3
+
+      - run: sed -E 's:RepositoryCommit>.+<\/:RepositoryCommit>${{ github.sha }}<\/:g' ButterCMS/ButterCMS.csproj > tmp && mv tmp ButterCMS/ButterCMS.csproj
```
- You don't need to escape `/` character (`\/` in `RepositoryCommit>.+<\/`) if you used `:` for the delimiter.
- You can edit files in-place using `-I` option.
```suggestion
      - run: sed -E -I'.bak' 's:RepositoryCommit>.+</:RepositoryCommit>${{ github.sha }}</:g' ButterCMS/ButterCMS.csproj ButterCMS/ButterCMS.csproj
```

vlnevyhosteny replied:
> You don't need to escape / character (\/ in RepositoryCommit>.+<\/) if you used : for the delimiter.

@Toreno96  I didn't work without escape on Mac 😢 . But it doesn't matter since in CI.

Toreno96 replied:
Weird, it worked on my Mac 😕 But yeah, that will run inside CI/CD, so it's not an issue. For the future, you could test using [gsed](https://formulae.brew.sh/formula/gnu-sed) or run Ubuntu in Docker if something's not working on your Mac.

Toreno96 commented on .github/workflows/nuget-publish.yml:
```diff
@@ -0,0 +1,70 @@
+        run: dotnet tool update Meziantou.Framework.NuGetPackageValidation.Tool --global
+
+      - name: Validate nuget package
+        run: meziantou.validate-nuget-package $(ls ${{ env.NUGET_DIR }}/*.nupkg | head -n 1) --excluded-rules Symbols
```
1. What are possible files that `ls ${{ env.NUGET_DIR }}/*.nupkg` will find?
2. Are we sure that `ls ${{ env.NUGET_DIR }}/*.nupkg | head -n 1` (the first file in the list) will _always_ evaluate to the exact file we want? IIRC it will be the first file in the lexicographical order.
3. [shellcheck](https://www.shellcheck.net/) finds a few issues with the code:

```
$ shellcheck -s bash a.sh

In a.sh line 1:
meziantou.validate-nuget-package $(ls ${{ env.NUGET_DIR }}/*.nupkg | head -n 1) --excluded-rules Symbols
                                 ^-- SC2046 (warning): Quote this to prevent word splitting.
                                   ^-----------------------------^ SC2012 (info): Use find instead of ls to better handle non-alphanumeric filenames.
                                      ^-----------------^ SC2086 (info): Double quote to prevent globbing and word splitting.
                                        ^--------------^ SC2296 (error): Parameter expansions can't start with {. Double check syntax.
                                                         ^-- SC1083 (warning): This } is literal. Check expression (missing ;/\n?) or quote it.

Did you mean:
meziantou.validate-nuget-package $(ls "${{ env.NUGET_DIR }"}/*.nupkg | head -n 1) --excluded-rules Symbols

For more information:
  https://www.shellcheck.net/wiki/SC2296 -- Parameter expansions can't start ...
  https://www.shellcheck.net/wiki/SC1083 -- This } is literal. Check expressi...
  https://www.shellcheck.net/wiki/SC2046 -- Quote this to prevent word splitt...
```

You may consider fixing them (excluding the `SC2296` one, which I believe is caused by double `{` in `${{ env.NUGET_DIR }}`, which is a valid syntax for variable substitution in GH actions).

vlnevyhosteny replied:
@Toreno96 There should be only one file in `${{ env.NUGET_DIR }}/*.nupkg` unless someone change packing process.  However, I changed the line to 

```bash
for FILE in ${{ env.NUGET_DIR }}/*.nupkg; do meziantou.validate-nuget-package "$FILE" --excluded-rules Symbols; done
```

so no file is going to be ignored. The same for the push. 

Also checked commands with **shellcheck**.

Toreno96 replied:
> There should be only one file

OK, but what's the example name of this file?

Toreno96 replied:
Cannot we simplify to:
```
$ meziantou.validate-nuget-package ${{ env.NUGET_DIR }}/*.nupkg --excluded-rules Symbols
```
??

Does `meziantou.validate-nuget-package` not accept multiple files at once?

vlnevyhosteny replied:
> > There should be only one file
> 
> OK, but what's the example name of this file?

`ButterCMS.2.0.0.nupkg`

vlnevyhosteny replied:
> Cannot we simplify to:
> 
> ```
> $ meziantou.validate-nuget-package ${{ env.NUGET_DIR }}/*.nupkg --excluded-rules Symbols
> ```
> 
> ??
> 
> Does `meziantou.validate-nuget-package` not accept multiple files at once?

Resolved in https://github.com/ButterCMS/buttercms-csharp/pull/40/commits/302382eee44defeb37f981f10c1f22263c6464d1

Toreno96 commented on .github/workflows/nuget-publish.yml:
```diff
@@ -0,0 +1,70 @@
+        uses: actions/setup-dotnet@v3
+
+      - name: Publish nuget package
+        run: dotnet nuget push $(ls ${{ env.NUGET_DIR }}/*.nupkg | head -n 1) --api-key "${{ secrets.NUGET_APIKEY }}" --source https://api.nuget.org/v3/index.json
```
shellcheck:
```

In a.sh line 1:
dotnet nuget push $(ls ${{ env.NUGET_DIR }}/*.nupkg | head -n 1) --api-key "${{ secrets.NUGET_APIKEY }}" --source https://api.nuget.org/v3/index.json
                  ^-- SC2046 (warning): Quote this to prevent word splitting.
                    ^-----------------------------^ SC2012 (info): Use find instead of ls to better handle non-alphanumeric filenames.
                       ^-----------------^ SC2086 (info): Double quote to prevent globbing and word splitting.
                         ^--------------^ SC2296 (error): Parameter expansions can't start with {. Double check syntax.
                                          ^-- SC1083 (warning): This } is literal. Check expression (missing ;/\n?) or quote it.
                                                                              ^---------------------^ SC2296 (error): Parameter expansions can't start with {. Double check syntax.

Did you mean: 
dotnet nuget push $(ls "${{ env.NUGET_DIR }"}/*.nupkg | head -n 1) --api-key "${{ secrets.NUGET_APIKEY }}" --source https://api.nuget.org/v3/index.json

For more information:
  https://www.shellcheck.net/wiki/SC2296 -- Parameter expansions can't start ...
  https://www.shellcheck.net/wiki/SC1083 -- This } is literal. Check expressi...
  https://www.shellcheck.net/wiki/SC2046 -- Quote this to prevent word splitt...
```

vlnevyhosteny replied:
Resolved in https://github.com/ButterCMS/buttercms-csharp/pull/40/commits/f3d2d5e3f00c7bfcdd3405cd42a7cfb56b556a32

Toreno96 replied:
`dotnet nuget push` seem to support pushing multiple files at once:
> Push all _.nupkg_ files in the current directory to the default push source specified in the NuGet config file:
> ```
> dotnet nuget push "*.nupkg"
> ```
> —https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-nuget-push

Therefore, that for-loop may be redundant.

Toreno96 replied:
One last thing 😅 
The docs specify that:
```
dotnet nuget push "*.nupkg"
```
> Note
> The enclosing quotes are required for shells such as bash that perform file globbing. For more information, see [NuGet/Home#4393](https://github.com/NuGet/Home/issues/4393#issuecomment-667618120).

So I think we need to do:
```
dotnet nuget push "${{ env.NUGET_DIR }}/*.nupkg" …
```

(I didn't test that, tho)

vlnevyhosteny replied:
@Toreno96 tested and resolved in https://github.com/ButterCMS/buttercms-csharp/pull/40/commits/23ec3acc174e05d1bf6516ee3a68016387ab7211

Toreno96 commented on .github/workflows/nuget-publish.yml:
```diff
@@ -0,0 +1,70 @@
+
+      - run: sed -E 's:RepositoryCommit>.+<\/:RepositoryCommit>${{ github.sha }}<\/:g' ButterCMS/ButterCMS.csproj > tmp && mv tmp ButterCMS/ButterCMS.csproj
+
+      - run: dotnet pack --configuration Release --output ${{ env.NUGET_DIR }}
```
shellcheck:
```

In a.sh line 1:
dotnet pack --configuration Release --output ${{ env.NUGET_DIR }}
                                             ^-----------------^ SC2086 (info): Double quote to prevent globbing and word splitting.
                                               ^--------------^ SC2296 (error): Parameter expansions can't start with {. Double check syntax.
                                                                ^-- SC1083 (warning): This } is literal. Check expression (missing ;/\n?) or quote it.

Did you mean: 
dotnet pack --configuration Release --output "${{ env.NUGET_DIR }"}

For more information:
  https://www.shellcheck.net/wiki/SC2296 -- Parameter expansions can't start ...
  https://www.shellcheck.net/wiki/SC1083 -- This } is literal. Check expressi...
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
```

vlnevyhosteny replied:
Resolved in https://github.com/ButterCMS/buttercms-csharp/pull/40/commits/f3d2d5e3f00c7bfcdd3405cd42a7cfb56b556a32

Toreno96 commented on .github/workflows/nuget-publish.yml:
```diff
@@ -0,0 +1,70 @@
+
+  pack: 
+    if: contains(github.event.head_commit.message, 'chore(release)') 
+    runs-on: ubuntu-latest
```
Let's pin this and all other `ubuntu`s to the specific version, similar to https://github.com/ButterCMS/buttercms-ruby/pull/28/files#diff-2c84033033d49186c63e6adcd705f63b11ae6814cd76c152c9c486d389fbccf3R23, to avoid unexpected failure in CI/CD when new Ubuntu is released with breaking changes.

vlnevyhosteny replied:
Resolved in https://github.com/ButterCMS/buttercms-csharp/pull/40/commits/f3d2d5e3f00c7bfcdd3405cd42a7cfb56b556a32

Toreno96 commented on .github/workflows/nuget-publish.yml:
```diff
@@ -0,0 +1,70 @@
+    if: contains(github.event.head_commit.message, 'chore(release)') 
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
```
Let's pin `actions/checkout` to more specific version, similar to https://github.com/ButterCMS/buttercms-ruby/pull/28/files#diff-1ac2730f9bc3ac91b4137899948628a3b5c6b181763cbfb93fa9d979bb6c8ac5R16.

vlnevyhosteny replied:
Resolved in https://github.com/ButterCMS/buttercms-csharp/pull/40/commits/f3d2d5e3f00c7bfcdd3405cd42a7cfb56b556a32

Toreno96 commented on .github/workflows/nuget-publish.yml:
```diff
@@ -0,0 +1,70 @@
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Setup .NET Core
+        uses: actions/setup-dotnet@v3
```
Can we pin dotnet to specific version, similar to https://github.com/ButterCMS/buttercms-ruby/pull/28/files#diff-1ac2730f9bc3ac91b4137899948628a3b5c6b181763cbfb93fa9d979bb6c8ac5R17-R19?

vlnevyhosteny replied:
Resolved in https://github.com/ButterCMS/buttercms-csharp/pull/40/commits/f3d2d5e3f00c7bfcdd3405cd42a7cfb56b556a32

Toreno96 commented on .github/workflows/nuget-publish.yml:
```diff
@@ -0,0 +1,70 @@
+
+      - run: dotnet pack --configuration Release --output ${{ env.NUGET_DIR }}
+
+      - uses: actions/upload-artifact@v3
```
Let's pin `actions/upload-artifact` to as specific version as possible.

vlnevyhosteny replied:
Resolved in https://github.com/ButterCMS/buttercms-csharp/pull/40/commits/f3d2d5e3f00c7bfcdd3405cd42a7cfb56b556a32

Toreno96 commented on .github/workflows/nuget-publish.yml:
```diff
@@ -0,0 +1,70 @@
+      - name: Setup .NET Core
+        uses: actions/setup-dotnet@v3
+
+      - uses: actions/download-artifact@v3
```
Let's pin to as specific version as possible.

vlnevyhosteny replied:
Resolved in https://github.com/ButterCMS/buttercms-csharp/pull/40/commits/f3d2d5e3f00c7bfcdd3405cd42a7cfb56b556a32

jakelumetta commented:
I've added `NUGET_APIKEY` as a secret to the repo.

Toreno96 reviewed (approved):
LGTM!
